### PR TITLE
[compiler-v2] Reorganize experiments and tests

### PR DESCRIPTION
**Pull request by wrwg**
This refactors the experiment system and our unit and transactional test framework.

- Experiments are now defined over a set of declarations which provide defaults. They can be set as 'on' and 'off', both over the command line, env, and programmatically.
- There is now an experiment for nearly each individual phase. Some experiments like `OPTIMIZE` also allow to turn on or off a group of other ones via an inheritance mechanism.
- There is a single function to create the env and the bytecode pipelines which considers all experiments, but ensures the same order of composition. This ends the ad-hoc way of how we defined pipelines in `testsuite.rs`, making sure we are testing the same as what we do in production. With this we cannot currently test different orderings, which we can get back too later if needed.
- The unit tests are now based on a set of test configurations provided via a table. Multiple configurations can apply to the same test file, allowing us to test different combinations (e.g. optimization on or off). Configurations can share the same .exp file or have individual ones via different endings.
- The datatest facility is used via a lower-level API, allowing to construct a single test case for cargo test/nextest for each (config, file) combination. This ensures that all those test can run in full concurrency as well as have precise error reporting.
- Something similar has been for transactional tests, which run now in different configurations.

Not all possible or wanted combinations have been explored, however, things should be in state such that we can now easily add more.

There is one caveat: `cargo nextest` is signifcantly slower then `cargo test` to execute the new test setup, since computation of the `datatest_stable` tests takes longer and is run by nextest many times. However, it is still in the acceptable range.

### CODE REVIEW

**Comment by brmataptos** [comment]
You need to update `aptos-move/aptos-transactional-test-harness/tests/v2-tests/default_int_size.exp` to pass tests.  It looks fine, just an expected failure, but an offset has apparently changed.

File: third_party/move/move-compiler-v2/src/experiments.rs
Comment (brmataptos) [inline third_party/move/move-compiler-v2/src/experiments.rs:46]: Does an "Experiment" represent a single experimental case vs the default?

If so, the description should say "Turns **off** a group of context checks", as that is the experiment condition.  Saying "on or off" leaves me confused about the experimental condition.  Same for the below.

Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/experiments.rs:46]: We also may want to have the names reflect what's happening in the experiment condition, e.g., "NO_CHECKS".

**Review by brmataptos — COMMENTED** [review]
Still reading, hopefully these are not already obsolete.

File: third_party/move/move-compiler-v2/src/experiments.rs
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/experiments.rs:46]: Oh, these are now "flags", not really experiments.  names matter.

**Review by vineethk — COMMENTED** [review]
Couple of high level points:
1. If we want to add a particular sequence of stackless bytecode pipeline (which is say quite different from what is in `lib.rs`), then we cannot do it with this new framework (at least, not yet, because we can only enable/disable stuff from `lib.rs`). Maybe this is okay for now, but might be worth thinking about how we could evolve this?
2. Before merging this code, we should make sure to remove old `.exp` files that *have not changed, but have now become irrelevant*.

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs
Comment (brmataptos) [inline third_party/move/testing-infra/transactional-test-runner/src/framework.rs:707]: You can avoid `mut options` by changing this to:
```
let options = experiments.iter().fold(options, 
        |options, (exp, value)| options.set_experiment(exp, value));
```
though it doesn't look clearer now that I've written it out.

Oh, well.

**Comment by wrwg** [comment]
> You need to update `aptos-move/aptos-transactional-test-harness/tests/v2-tests/default_int_size.exp` to pass tests. It looks fine, just an expected failure, but an offset has apparently changed.

Thanks, fixed

File: third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs
Comment (brmataptos) [inline third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs:91]: Sorry for being stupid, but what does `main` do in this context?  Why didn't we need it before?  Feel free to just tell me the relevant crate name to go find docs.

**Review by brmataptos — COMMENTED** [review]
We really need to rename "experiment" to something else, though I'm not sure what.  "enable_pass"?  "use_pass"?  Perhaps a different tense even.  Please file an issue if you don't rename it here.

I wonder if we might also run into issues with wanting to try reordering passes in the future, but we don't have to boil the ocean here.

File: third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs:91]: The `main` was generated before by the `datatest_stable::harness!(run_test, TEST_DIR, r".*\.(mvir|move)$")` macro (example of use). Now we are doing it manually and build up the list of tests ourselves. `main` is really the entrypoint for tesing here.

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs
Comment (wrwg) [reply third_party/move/testing-infra/transactional-test-runner/src/framework.rs:707]: Yea, sometimes I really prefer the good old imperative loop...

**Comment by wrwg** [comment]
> We really need to rename "experiment" to something else, though I'm not sure what. "enable_pass"? "use_pass"? Perhaps a different tense even. Please file an issue if you don't rename it here.
> 
> I wonder if we might also run into issues with wanting to try reordering passes in the future, but we don't have to boil the ocean here.

I agree, lets open a bug for renaming this to `flags` or similar.

The reordering issue I mentioned also in the PR description. There are pros and cons. For now, testing the same order as we actually run in production seems to more on the plus side. We can come back to re-ordering later.

File: third_party/move/move-compiler-v2/src/lib.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/lib.rs]: Was already remove 1-2 commits ago, please refresh.

File: third_party/move/move-compiler-v2/src/lib.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/lib.rs:341]: Was already removed.

File: third_party/move/move-compiler-v2/src/lib.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/lib.rs:374]: The v1 compiler leaves specs alone. Added a comment.

File: third_party/move/move-compiler-v2/src/lib.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/lib.rs:412]: Thats why its the last step in the checking pipeline.

File: third_party/move/move-compiler-v2/tests/testsuite.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/testsuite.rs:77]: Done

File: third_party/move/move-compiler-v2/tests/testsuite.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/testsuite.rs:461]: I took that over as is, but fixed.

File: third_party/move/move-compiler-v2/tests/testsuite.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/testsuite.rs:42]: This is a standard concept, but done.

File: third_party/move/move-model/bytecode/src/function_target_pipeline.rs
Comment (wrwg) [reply third_party/move/move-model/bytecode/src/function_target_pipeline.rs:310]: I leave this to you. Perhaps we need it later? Not sure.

File: third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs:66]: I'm not sure what you mean. Cloned is needed here.

**Comment by wrwg** [comment]
> Reference

**Comment by wrwg** [comment]
> Couple of high level points:
> 
> 1. If we want to add a particular sequence of stackless bytecode pipeline (which is say quite different from what is in `lib.rs`), then we cannot do it with this new framework (at least, not yet, because we can only enable/disable stuff from `lib.rs`). Maybe this is okay for now, but might be worth thinking about how we could evolve this?
> 2. Before merging this code, we should make sure to remove old `.exp` files that _have not changed, but have now become irrelevant_.


Can I please encourage everybody to read PR descriptions? The re-ordering is tackled there, and I think its a much bigger advantage that we actually test the pipeline we use in production. We can come back to the re-ordering if there is actually a use case. It is not the case we tested re-ordering before, it was just completely chaotic and diverging from reality as more phases were added.

Regards (2) I think I took care of it.

File: third_party/move/move-compiler-v2/src/experiments.rs
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/experiments.rs:46]: Experiments have values as well, so this wouldn't be the reason why to change the name. In fact at Google various experiment frameworks have even more then boolean, there can be enums etc. (Look at the experiments of Chrome.)

If you'd read the PR description this confusing could have been avoided, that they are on/off now is prominently documented there, also in `experiments.rs` ;-)

File: third_party/move/move-model/bytecode/src/function_target_pipeline.rs
Comment (vineethk) [reply third_party/move/move-model/bytecode/src/function_target_pipeline.rs:310]: Sounds good.

File: third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs
Comment (vineethk) [reply third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs:66]: Sorry, it applies to one line later (typo fix confio -> config). I think the line numbers got messed up because of an intermediate push or something.

**Comment by vineethk** [comment]
> Can I please encourage everybody to read PR descriptions? The re-ordering is tackled there, and I think its a much bigger advantage that we actually test the pipeline we use in production. We can come back to the re-ordering if there is actually a use case. It is not the case we tested re-ordering before, it was just completely chaotic and diverging from reality as more phases were added.
> 
> Regards (2) I think I took care of it.

I did read it, but forgot after reading a bunch of code. You can consider that comment resolved.

Regards (2), I misunderstood the presence of some `exp` files, it is legitimate for them to be present. So this can also be considered resolved.

**Review by vineethk — APPROVED** [review]

**Review by brmataptos — COMMENTED** [review]
I don't remember what I said below.

File: third_party/move/move-compiler-v2/tests/file-format-generator/fields.opt.exp
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/file-format-generator/fields.opt.exp:55]: I've filed https://github.com/aptos-labs/aptos-core/issues/12648 to look into it further.

File: third_party/move/move-prover/tests/sources/regression/type_param_bug_200228.v2_exp
Comment (brmataptos) [reply third_party/move/move-prover/tests/sources/regression/type_param_bug_200228.v2_exp:12]: Filed issue https://github.com/aptos-labs/aptos-core/issues/12649 to track.

**Review by brmataptos — APPROVED** [review]

**Comment by wrwg** [comment]
This is actually not a stack trace but an execution trace (i.e. the sequence of steps in the program execution leading to the verification failure). With slight changes in the bytecode this can change, and an entry without any variables is possible, because which variables are printed in each step is based on some heuristics of what state has changed. The entire logic around producing this kind of execution trace is super complex piece of code in the prover as the SMT counter example is indeed completely different than the original program. So this is kind of expected. Thanks for flagging, but unless you want to discuss more, I suggest to close.

**Pull request merged**
